### ISSUE TITLE
Deberta's Enhanced Masked Decoder

### ISSUE THREAD

**Issue by alexvaca0** [huggingface/transformers#14585]
# 🚀 Feature request

I don't know if this is a feature request or the feature is already included but I wasn't able to find it. Reviewing Deberta's code, I tried to look for the Enhanced Masked Decoder, which is supposed to take into account absolute positional embeddings. However, when looking both DebertaModel and DebertaForMLM (and its submodules) I don't see where this is done... Could someone please tell me, if the Enhanced Masked Decoder is applied somewhere? Reading the Deberta paper it is clear this is one of the key improvements with respect to BERT and RoBERTa, and I think the Transformers' implementation of Deberta was carried out by the original author of the model, therefore I am sure this part is not missing and it's me that I don't find it... If someone can help me with this I'd really appreciate it :) 

@patrickvonplaten @sgugger @BigBird01

**Comment by BigBird01** [comment]
The implementation of Enhanced Mask Decoder was here, https://github.com/microsoft/DeBERTa/blob/master/DeBERTa/apps/models/masked_language_model.py
We will port the implementation to transformers.

**Comment by alexvaca0** [comment]
And why did you decide to leave it out from the Transformers library when implementing Deberta in Transformers? 
If I see the model is implemented by the original author I expect that the implementation is complete and fully reliable... 
At my company we have spent huge amounts of money for training a Deberta model in Spanish, and now, what can we do? Is it possible to continue training with this new update with no problems of convergence (As we are slightly changing the network and the optimizer states belonged to another network)?
Actually this is very disappointing, we chose Deberta because of the proven performance in the paper, and because we saw that Transformers' Deberta model was developed by the original author, therefore we expected to be using the exact same Deberta model that was developed for the paper. We need to solve this, as we will continue training from step 250k and we want the actual and original Deberta model.
If there is anything me, or any other member of the community can do to save you some time and help out with this, just say it, or at least document that it is missing; if I had seen in the documentation that the model was incomplete I couldn't be angry or frustrated, but I had to dig very deeply into both the code and the paper to find this... to find it almost a year after the Deberta modeling code is pushed into transformers, and I think that is neither serious nor rigorous.
I insist, please ask for help if you need it or if you don't have the time to look into it, and sorry if my tone sounds angry, but please understand that discovering this after so much investment in time and money in our model is a frustrating inconvenience we didn't expect given it was developed by the original author of the model.

**Comment by BigBird01** [comment]
Hi, Alex, thanks for you interesting to our work and sorry for the inconvenience.  The missing EMD part will not affect the convergence of your training. It will slightly affect the perplexity of MLM training. The inconsistent issue between transformers and the official implementation is a bug due to history reason. We will fix it and improve our future practice to maintain higher quality bars at both side. If you have other issue with pre-training, please feel free to reach me by mail directly. Thanks!

**Comment by alexvaca0** [comment]
Thank you so much for the understanding and for the kind reply. Of course I had interest in your work from the moment I first read the paper, it's so original and your results are impressive. I really appreciate you taking the time to fix this, I'll be looking forward to try the changes. From what you say, do you think that we could safely continue training from the last checkpoint with the new model updates without trouble? Or would we need to make an adjustment to make things work? 
Again, thanks for the kind explanation, and for sharing your work with the community.

**Comment by BigBird01** [comment]
Thanks for you understanding and kind words, Alex:) 
I think you can just continue train with your current settings as long as the training loss didn't go wild. 
Basically, except for DA and EMD described in our paper, there are a few other things important for pre-training.

1. Batch size, learning rate, warmup. For base model, our best local settings are: 2k batch size, 1M steps with learning rate 3e-4 and warmup for 10k steps. You may also need to set beta2 of Adam to 0.98 to stabilize the training with large batch size.
2. Model size
3. Vocabulary
4. Data volume and quality


We just added a [document](https://github.com/microsoft/DeBERTa/tree/master/experiments/language_model) on pre-training in our official repo, you make take it as a reference for your settings. And you are welcome to leave comments on it.
